File: server/lib/format.js
'use strict';

const ANSI = /\x1b\[[0-9;?]*[ -/]*[@-~]/g;

function toAscii(s) {
  return String(s)
    .split('\t').join('  ')
    .replace(/[‘’‛]/g, "'")
    .replace(/[“”]/g, '"')
    .replace(/[–—−]/g, '-')
    .replace(/…/g, '...')
    .replace(/[•·]/g, '*')
    .replace(/[❯▶›]/g, '>')
    .replace(/[→⇒]/g, '->')
    .replace(/[←⇐]/g, '<-')
    .replace(/[✓✔]/g, 'v')
    .replace(/ /g, ' ')
    .replace(/[^\x0A\x20-\x7E]/g, '');
}

function wrapLine(line, cols) {
  if (line.length <= cols) return [line];
  const out = [];
  const words = line.split(/\s+/);
  let cur = '';
  for (let word of words) {
    while (word.length > cols) {
      if (cur) { out.push(cur); cur = ''; }
      out.push(word.slice(0, cols));
      word = word.slice(cols);
    }
    if (!cur) { cur = word; continue; }
    if (cur.length + 1 + word.length <= cols) { cur += ' ' + word; continue; }
    out.push(cur);
    cur = word;
  }
  if (cur) out.push(cur);
  return out.length ? out : [''];
}

function sliceIntoCards(text, cols, linesPerCard, maxCards = 40) {
  const cleaned = toAscii(String(text).replace(ANSI, '').replace(/\r/g, ''));
  let rawLines = cleaned.split('\n');
  while (rawLines.length && rawLines[rawLines.length - 1].trim() === '') rawLines.pop();

  const wrapped = [];
  for (const line of rawLines) {
    for (const w of wrapLine(line, cols)) {
      wrapped.push(w.replace(/\s+$/, ''));
    }
  }

  const cards = [];
  for (let i = 0; i < wrapped.length; i += linesPerCard) {
    cards.push(wrapped.slice(i, i + linesPerCard));
  }
  if (cards.length > maxCards) return cards.slice(cards.length - maxCards);
  return cards;
}

module.exports = { toAscii, wrapLine, sliceIntoCards };

File: server-go/internal/screen/format.go
package screen

import "strings"

var translit = map[rune]string{
	'‘': "'", '’': "'", '‛': "'",
	'“': "\"", '”': "\"",
	'–': "-", '—': "-", '−': "-",
	'…': "...",
	'•': "*", '·': "*",
	'❯': ">", '▶': ">", '›': ">",
	'→': "->", '⇒': "->",
	'←': "<-", '⇐': "<-",
	'✓': "v", '✔': "v",
	' ': " ",
}

func ToAscii(s string) string {
	s = strings.ReplaceAll(s, "\t", "  ")
	var b strings.Builder
	for _, r := range s {
		if rep, ok := translit[r]; ok {
			b.WriteString(rep)
			continue
		}
		if r == '\n' || (r >= 0x20 && r <= 0x7e) {
			b.WriteRune(r)
		}
	}
	return b.String()
}

func WrapLine(line string, cols int) []string {
	runes := []rune(line)
	if len(runes) <= cols {
		return []string{line}
	}
	out := []string{}
	cur := ""
	for _, word := range strings.Fields(line) {
		wr := []rune(word)
		for len(wr) > cols {
			if cur != "" {
				out = append(out, cur)
				cur = ""
			}
			out = append(out, string(wr[:cols]))
			wr = wr[cols:]
		}
		word = string(wr)
		if cur == "" {
			cur = word
			continue
		}
		if len([]rune(cur))+1+len(wr) <= cols {
			cur += " " + word
			continue
		}
		out = append(out, cur)
		cur = word
	}
	if cur != "" {
		out = append(out, cur)
	}
	if len(out) == 0 {
		return []string{""}
	}
	return out
}

func SliceIntoCards(text string, cols, linesPerCard, maxCards int) [][]string {
	cleaned := ToAscii(strings.ReplaceAll(StripAnsi(text), "\r", ""))
	rawLines := strings.Split(cleaned, "\n")
	for len(rawLines) > 0 && strings.TrimSpace(rawLines[len(rawLines)-1]) == "" {
		rawLines = rawLines[:len(rawLines)-1]
	}

	wrapped := []string{}
	for _, line := range rawLines {
		for _, w := range WrapLine(line, cols) {
			wrapped = append(wrapped, strings.TrimRight(w, " "))
		}
	}

	cards := [][]string{}
	page := max(1, linesPerCard)
	for i := 0; i < len(wrapped); i += page {
		end := min(i+page, len(wrapped))
		cards = append(cards, wrapped[i:end])
	}
	if maxCards > 0 && len(cards) > maxCards {
		return cards[len(cards)-maxCards:]
	}
	return cards
}
